projects/plugins/crm/src/automation/commons/triggers/contacts/class-contact-deleted.php

```diff
@@ -0,0 +1,49 @@
+<?php
+/**
+ * Jetpack CRM Automation Contact_Deleted trigger.
+ *
+ * @package automattic/jetpack-crm
+ */
+
+namespace Automattic\Jetpack\CRM\Automation\Triggers;
+
+use Automattic\Jetpack\CRM\Automation\Automation_Exception;
+use Automattic\Jetpack\CRM\Automation\Automation_Workflow;
+use Automattic\Jetpack\CRM\Automation\Base_Trigger;
+
+/**
+ * Adds the Contact_Deleted class.
+ */
+class Contact_Deleted extends Base_Trigger {
+
+	/**
+	 * Contructs the Contact_Deleted instance.
+	 */
+	public function __construct() {
+		$trigger_data = array(
+			'name'        => 'contact_delete',
+			'title'       => 'Contact Deleted',
+			'description' => 'Triggered when a contact is deleted.',
+			'category'    => 'contact',
+		);
+
+		parent::__construct( $trigger_data );
+	}
+
+	/**
+	 * Init the trigger. Listen to the desired event
+	 *
+	 * @param Automation_Workflow $workflow The workflow to which the trigger belongs.
+	 * @throws Automation_Exception Throws a 'class not found' or general error.
+	 */
+	public function init( Automation_Workflow $workflow ) {
+		add_action(
+			'jpcrm_automation_contact_delete',
+			function ( $contact_data ) use ( $workflow ) {
+				$workflow->execute( $this, $contact_data );
+			},
+			10,
+			2
+		);
+	}
```

Comment:
Could this be simplified in a more traditional WP way by creating a callback method as part of the trigger class instead of relying on parents scope of an anonymous function? 😄 

AFAIK then we could achieve the same by doing something like this: 

```suggestion
	public function init( Automation_Workflow $workflow ) {
		add_action( 'jpcrm_automation_contact_delete', array( $workflow, 'callback' ), 10 );
	}

	public function callback( $contact_data ) {
		// do something with the contact data
	}
```

That way `init()` is only responsible for triggering actions (read: separation of concerns could become clearer this way)

Reply:
> That way init() is only responsible for triggering actions

Here, the init method is not triggering the action, it's adding the action hook to the WP, and then the callback function (embedded function, using closures way), that executes the workflow. 

If we want to extract the closure, it should be:

```
	private $workflow;
	
	public function init( Automation_Workflow $workflow ) {
		$this->workflow = $workflow;
		add_action( 'jpcrm_automation_contact_delete', array( $workflow, 'execute_workflow' ) );
	}
	
	public function execute_workflow( $contact_data ) {
		if ( $this->workflow ) {
			$this->workflow->execute( $this, $contact_data );
		}
	}
```

The `do_action`, in this case, will have 1 argument, the contact object/data. In this case, the $workflow should be stored in the trigger instance (private property). I'm fine extracting it this way, but I have to update `Base_Trigger`, so triggers implementations don't have to be aware of it

Reply:
I, personally, find it more like "the WordPress way" to do it the way you just described 😄

Reply:
Karen and me tested this solution, and it works well and easier to test it using the PHPUnit Mock builder